src/translate.c

```diff
@@ -110,87 +113,90 @@ static r_ssize chr_find_normalize_start(SEXP x, r_ssize size) {
 
 // -----------------------------------------------------------------------------
 
-static SEXP list_normalize_encoding(SEXP x, r_ssize size, r_ssize start) {
-  x = PROTECT(r_clone_referenced(x));
-  const SEXP* p_x = VECTOR_PTR_RO(x);
+static
+SEXP list_normalize_encoding(SEXP x) {
+  int nprot = 0;
 
-  for (r_ssize i = start; i < size; ++i) {
-    SEXP elt = p_x[i];
-
-    switch (TYPEOF(elt)) {
-    case STRSXP: {
-      r_ssize elt_size = r_length(elt);
-      r_ssize elt_start = chr_find_normalize_start(elt, elt_size);
+  r_ssize size = r_length(x);
+  const SEXP* p_x = VECTOR_PTR_RO(x);
 
-      if (elt_size == elt_start) {
-        break;
-      }
+  for (r_ssize i = 0; i < size; ++i) {
+    SEXP elt_old = p_x[i];
 
-      elt = chr_normalize_encoding(elt, elt_size, elt_start);
-      SET_VECTOR_ELT(x, i, elt);
-      break;
+    SEXP elt_new = obj_normalize_encoding(elt_old);
+    if (elt_old == elt_new) {
+      continue;
     }
-    case VECSXP: {
-      r_ssize elt_size = r_length(elt);
-      r_ssize elt_start = list_find_normalize_start(elt, elt_size);
-
-      if (elt_size == elt_start) {
-        break;
-      }
+    PROTECT(elt_new);
 
-      elt = list_normalize_encoding(elt, elt_size, elt_start);
-      SET_VECTOR_ELT(x, i, elt);
-      break;
-    }
-    default:
-      continue;
+    if (MAYBE_REFERENCED(x)) {
```

Comment:
Mention that this can only be cloned once, at which point `x` is free of references.